contracts/galactic-express/src/lib.rs

```diff
@@ -117,57 +137,155 @@ impl Contract {
+            }
+
+            if participants.len() >= PARTICIPANTS - 1 {
+                return Err(Error::SessionFull);
+            }
+
+            participant.check()?;
+            participants.insert(msg_source, participant.clone());
+            self.player_to_game_id.insert(msg_source, creator);
+
+            Ok(Event::Registered(msg_source, participant))
+        } else {
+            Err(Error::NoSuchGame)
         }
+    }
 
-        participant.check()?;
-        participants.insert(msg_source, participant);
+    fn cancel_register(&mut self) -> Result<Event, Error> {
+        let msg_source = msg::source();
 
-        Ok(Event::Registered(msg_source, participant))
+        let creator = self
+            .player_to_game_id
+            .get(&msg_source)
+            .ok_or(Error::Unregistered)?;
+        let game = self.games.get_mut(creator).ok_or(Error::NoSuchGame)?;
+
+        if msg_source != game.admin {
+            let participants = game.stage.mut_participants()?;
+            if participants.contains_key(&msg_source) {
+                send_value(msg_source, game.bid);
+                participants.remove(&msg_source).expect("Critical error");
+                self.player_to_game_id.remove(&msg_source);
+            } else {
+                return Err(Error::NoSuchPlayer);
+            }
+            Ok(Event::CancelRegistration)
+        } else {
+            Err(Error::NotForAdmin)
+        }
     }
+    fn delete_player(&mut self, player_id: ActorId) -> Result<Event, Error> {
+        let msg_source = msg::source();
 
-    async fn start_game(&mut self, mut participant: Participant) -> Result<Event, Error> {
-        self.check_admin()?;
+        if let Some(game) = self.games.get_mut(&msg_source) {
+            if let Stage::Results(_) = game.stage {
+                return Err(Error::SessionEnded);
+            }
 
-        let participants = self.stage.mut_participants()?;
+            let participants = game.stage.mut_participants()?;
 
-        if participants.is_empty() {
-            return Err(Error::NotEnoughParticipants);
+            if participants.contains_key(&player_id) {
+                send_value(player_id, game.bid);
+                participants.remove(&player_id).expect("Critical error");
+                self.player_to_game_id.remove(&player_id);
+            } else {
+                return Err(Error::NoSuchPlayer);
+            }
+
+            Ok(Event::PlayerDeleted { player_id })
+        } else {
+            Err(Error::NoSuchGame)
         }
+    }
+
+    async fn start_game(&mut self, fuel_amount: u8, payload_amount: u8) -> Result<Event, Error> {
+        let msg_source = msg::source();
 
+        let game = self.games.get_mut(&msg_source).ok_or(Error::NoSuchGame)?;
+        let participant = Participant {
+            id: msg_source,
+            name: game.admin_name.clone(),
+            fuel_amount,
+            payload_amount,
+        };
         participant.check()?;
```

Comment:
This is wrong design. One shouldn't be allowed to create an instance in a broken state